File: model_1/icpc-jakarta-2018/binary_1_solution.py
import sys

def main():
    data = sys.stdin.read().splitlines()
    if not data: 
        return
    K = int(data[0].strip())
    S = data[1].strip()
    n = len(S)
    BIG = K + 1
    
    for L in range(n, 0, -1):
        if L == 1:
            print(n - 1)
            return
        else:
            dp_prev = [BIG] * n
            for i in range(n):
                if S[i] == '1':
                    dp_prev[i] = 1
                    
            for j in range(2, L + 1):
                dp_curr = [BIG] * n
                min_prev = BIG
                for i in range(n):
                    if min_prev != BIG:
                        bit_val = 1 if S[i] == '1' else 0
                        candidate = min_prev * 2 + bit_val
                        if candidate <= K:
                            dp_curr[i] = candidate
                    if dp_prev[i] != BIG and dp_prev[i] < min_prev:
                        min_prev = dp_prev[i]
                
                if all(x == BIG for x in dp_curr):
                    break
                dp_prev = dp_curr
            else:
                for value in dp_prev:
                    if value <= K:
                        print(n - L)
                        return
    print(n - 1)

if __name__ == "__main__":
    main()

File: model_1/icpc-jakarta-2018/binary_1_solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

typedef unsigned long long ull;

int main() {
    ull K;
    string S;
    cin >> K;
    cin >> S;
    int n = S.size();

    ull BIG = K + 1;

    for (int L = n; L >= 1; L--) {
        if (L == 1) {
            cout << n - 1 << endl;
            return 0;
        }

        vector<ull> dp_prev(n, BIG);
        for (int i = 0; i < n; i++) {
            if (S[i] == '0') {
                dp_prev[i] = 0;
            } else {
                dp_prev[i] = 1;
            }
        }

        for (int j = 2; j <= L; j++) {
            vector<ull> dp_curr(n, BIG);
            ull min_so_far = BIG;

            for (int i = 0; i < n; i++) {
                if (min_so_far != BIG) {
                    ull bit_val = (S[i] == '1') ? 1 : 0;
                    ull candidate = min_so_far * 2 + bit_val;
                    if (candidate <= K) {
                        dp_curr[i] = candidate;
                    }
                }

                if (dp_prev[i] != BIG && dp_prev[i] <= K) {
                    if (dp_prev[i] != 0) {
                        if (min_so_far == BIG) {
                            min_so_far = dp_prev[i];
                        } else {
                            if (dp_prev[i] < min_so_far) {
                                min_so_far = dp_prev[i];
                            }
                        }
                    }
                }
            }
            dp_prev = dp_curr;
        }

        for (ull val : dp_prev) {
            if (val <= K) {
                cout << n - L << endl;
                return 0;
            }
        }
    }

    cout << n - 1 << endl;
    return 0;
}